src/peer-store/book.js

```diff
@@ -0,0 +1,80 @@
+'use strict'
+
+const errcode = require('err-code')
+const PeerId = require('peer-id')
+const PeerInfo = require('peer-info')
+
+const {
+  ERR_INVALID_PARAMETERS
+} = require('../errors')
+
+/**
+ * The Book is the skeleton for the PeerStore books.
+ */
+class Book {
+  constructor (eventEmitter, eventName, eventProperty) {
+    this.eventEmitter = eventEmitter
+    this.eventName = eventName
+    this.eventProperty = eventProperty
+
+    /**
+     * Map known peers to their data.
+     * @type {Map<string, Array<Data>}
+     */
+    this.data = new Map()
+  }
+
+  /**
+   * Set known data of a provided peer.
+   * @param {PeerId} peerId
+   * @param {Array<Data>|Data} data
+   * @param {Object} [options]
+   * @param {boolean} [options.replace = true] wether data received replace stored the one or a unique union is performed.
+   */
+  set (peerId, data, options) {
+    throw errcode(new Error('set must be implemented by the subclass'), 'ERR_NOT_IMPLEMENTED')
+  }
+
+  /**
+   * Get the known data of a provided peer.
+   * @param {PeerId} peerId
+   * @returns {Array<Data>}
+   */
+  get (peerId) {
+    if (!PeerId.isPeerId(peerId)) {
+      throw errcode(new Error('peerId must be an instance of peer-id'), ERR_INVALID_PARAMETERS)
+    }
```

Comment:
I personally don't like to see `cannot access `toString()` of undefined as an error, as someone needs to go into the code/docs and see what is the issue. This way, I think the error is useful here. if we do not try to access properties of it, I would agree.

Reply:
We can table this discussion for outside of this PR as we're already doing similar checks elsewhere. In short, I'd rather see us move to support static type checking for users that want it instead of doing runtime checks. While they're not big, doing these type checks everywhere slowly adds to bundle size as well.

Reply:
Should we open a general issue to discuss this?